spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2017-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.bigquery.core;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.bigquery.storage.v1.AppendRowsResponse;
+import com.google.cloud.bigquery.storage.v1.BigQueryWriteClient;
+import com.google.cloud.bigquery.storage.v1.CreateWriteStreamRequest;
+import com.google.cloud.bigquery.storage.v1.Exceptions;
+import com.google.cloud.bigquery.storage.v1.Exceptions.StorageException;
+import com.google.cloud.bigquery.storage.v1.FinalizeWriteStreamResponse;
+import com.google.cloud.bigquery.storage.v1.JsonStreamWriter;
+import com.google.cloud.bigquery.storage.v1.TableName;
+import com.google.cloud.bigquery.storage.v1.WriteStream;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Descriptors.DescriptorValidationException;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Phaser;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.GuardedBy;
+import org.json.JSONArray;
+
+/** Helper class for using BigQuery storage write API in exactly once delivery mode */
+public class BigQueryJsonDataWriter {
+
+  private JsonStreamWriter streamWriter;
+  // Track the number of in-flight requests to wait for all responses before shutting down.
+  private final Phaser inflightRequestCount = new Phaser(1);
+
+  private final Object lock = new Object();
+
+  private final Logger logger = Logger.getLogger(this.getClass().getName());
+
+  @GuardedBy("lock")
+  private RuntimeException error = null;
+
+  void initialize(TableName parentTable, BigQueryWriteClient client)
```

Comment:
Is this object generally meant to be re-used between different table? From the usage in `BigQueryTemplate` it does not seem so. Would it make sense to make this object immutable, meant for one-time use against a single table? Right now, some very strange things would happen if `initialize()` is called the second time while the first write is in progress.

Reply:
Thanks for pointing this out, This is supposed to be initialised just once for the given table. I have added a parameterised constructor @ `BigQueryJsonDataWriter` to initialise class and have removed the `initialize()` method. Also, I have marked `streamWriter` as `final`